### ISSUE TITLE
Missing error on switch/case type mismatch

### ISSUE THREAD

**Issue by askeksa-google** [dart-lang/sdk#34207]
From `co19_2/Language/Statements/Switch/execution_t01`:
```dart
main() {
   switch (1) {
     case '1': break;
     case '2': break;
     default:
   }

  switch ('abc') {
    case 1: break;
  }
}
```
The spec says (in 17.9):
"It is a compile-time error if the type of e may not be assigned to the type of ek."

From `co19_2/Language/Statements/Switch/expressions_t01`:
```dart
main() {
  var x = 1;
  switch (x) {
    case 0:
    case 2:
    case "false":
  }
}
```
The spec says (in 17.9):
"it is a compile-time error if the expressions ek are not compile-time constants
for all k ∈ 1..n. It is a compile-time error if the values of the expressions ek are
not either:
• instances of the same class C, for all k ∈ 1..n, or
• instances of a class that implements int, for all k ∈ 1..n, or
• instances of a class that implements String, for all k ∈ 1..n.
In other words, all the expressions in the cases evaluate to constants of the
exact same user defined class or are of certain known types."

**Comment by dhil** [comment]
[CL 71981](https://dart-review.googlesource.com/c/sdk/+/71981) implements the assignability check.

**Comment by kmillikin** [comment]
Fixed in a56203041fa95d4d20926b3491018b61a9ab131d.

**Comment by kmillikin** [comment]
I should say, is this fixed in a56203041fa95d4d20926b3491018b61a9ab131d?

**Comment by dhil** [comment]
The check for whether the switch expression is assignable to each case expression has been implemented. But the second part of the issue has not been implemented yet:
> The spec says (in 17.9):
> "it is a compile-time error if the expressions ek are not compile-time constants
> for all k ∈ 1..n. It is a compile-time error if the values of the expressions ek are
> not either:
> • instances of the same class C, for all k ∈ 1..n, or
> • instances of a class that implements int, for all k ∈ 1..n, or
> • instances of a class that implements String, for all k ∈ 1..n.
> In other words, all the expressions in the cases evaluate to constants of the
> exact same user defined class or are of certain known types."

Partly because the language team is considering changing this part of the specification. But also because as specified now, we may need constant evaluation to implement it properly.

**Comment by srawlins** [comment]
The second part, regarding const, seems to be implemented. Dart VM now produces an error:

```
$ cat 34207.dart 
main() {
  switch ('abc') {
    case s: break;
  }
}

String get s => 'Hello';
$ dart 34207.dart
34207.dart:3:10: Error: Constant evaluation error:
    case s: break;
         ^
34207.dart:3:10: Context: The invocation of 's' is not allowed within a const context.
    case s: break;
         ^
34207.dart:3:5: Context: While analyzing:
    case s: break;
    ^
```
